File: backend/app/api/v1/research_lab.py
"""Simulator Research Lab V0 API."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from app.infrastructure.db.session import core_session
from app.modules.research_lab.application.compare import compare_runs
from app.modules.research_lab.application.service import (
    get_research_run,
    launch_research_run,
    list_research_runs,
    plan_quick_suite,
    research_options,
    run_quick_suite,
)
from app.modules.research_lab.errors import ResearchLabError

router = APIRouter()


class LaunchRequest(BaseModel):
    candidate_id: str = "prediction_ml_candidate/v0"
    segment: str = "DEVELOPMENT_OOS"
    policy_id: str
    risk_id: str
    commission_bps: float = 10.0
    slippage_bps: float = 0.0
    date_from: str | None = None
    date_to: str | None = None
    initial_capital: float = 1_000_000.0
    name: str | None = None
    note: str | None = None
    force_rerun: bool = False


class SuiteRequest(BaseModel):
    candidate_id: str = "prediction_ml_candidate/v0"
    date_from: str | None = None
    date_to: str | None = None
    initial_capital: float = 1_000_000.0


def _http_error(exc: ResearchLabError) -> HTTPException:
    status = 400
    if exc.code in {"RUN_NOT_FOUND"}:
        status = 404
    if exc.code in {"HOLDOUT_LAUNCH_FORBIDDEN"}:
        status = 403
    return HTTPException(status_code=status, detail=exc.to_dict())


@router.get("/options")
def api_options() -> dict:
    return research_options()


@router.get("/runs")
def api_list_runs(
    limit: int = Query(100, ge=1, le=200),
    policy_id: str | None = None,
    risk_id: str | None = None,
    status: str | None = None,
    segment: str | None = None,
    commission_bps: float | None = None,
    sort: str = Query("newest"),
) -> dict:
    with core_session() as session:
        items = list_research_runs(
            session,
            limit=limit,
            policy_id=policy_id,
            risk_id=risk_id,
            status=status,
            segment=segment,
            commission_bps=commission_bps,
            sort=sort,
        )
        return {"items": items}


@router.get("/runs/{run_id}")
def api_get_run(run_id: int) -> dict:
    try:
        with core_session() as session:
            return get_research_run(session, run_id)
    except ResearchLabError as exc:
        raise _http_error(exc) from exc


@router.post("/runs")
def api_launch_run(body: LaunchRequest) -> dict:
    try:
        with core_session() as session:
            out = launch_research_run(session, body.model_dump())
            session.commit()
            return out
    except ResearchLabError as exc:
        raise _http_error(exc) from exc


@router.get("/compare")
def api_compare(run_ids: str = Query(..., description="Comma-separated run ids")) -> dict:
    try:
        ids = [int(x.strip()) for x in run_ids.split(",") if x.strip()]
        with core_session() as session:
            return compare_runs(session, ids)
    except ResearchLabError as exc:
        raise _http_error(exc) from exc
    except ValueError as exc:
        raise HTTPException(
            status_code=400,
            detail={"code": "INVALID_RUN_IDS", "message": "Некорректный список run_ids"},
        ) from exc


@router.post("/suites/plan")
def api_suite_plan(body: SuiteRequest) -> dict:
    try:
        with core_session() as session:
            return plan_quick_suite(session, body.model_dump())
    except ResearchLabError as exc:
        raise _http_error(exc) from exc


@router.post("/suites")
def api_suite_run(body: SuiteRequest) -> dict:
    try:
        with core_session() as session:
            out = run_quick_suite(session, body.model_dump())
            session.commit()
            return out
    except ResearchLabError as exc:
        raise _http_error(exc) from exc

File: frontend/src/api/researchLab.ts
import { apiRequest, queryString } from "./client";
import type { SimulationMetrics, SimulationRunSummary, SimulationSpecSummary } from "./simulator";

/** Research Lab V0 — historical experiments over frozen Prediction Candidates. */

export interface CandidateOption {
  id: string;
  candidate_name: string;
  candidate_version: string;
  human_name: string;
  technical_line: string;
  research_verdict: string;
  model_type: string;
  target_label: string;
  /** EXPECTED_RETURN | RANKING_SCORE */
  prediction_semantic?: string;
  /** Human output label, e.g. «Рейтинговый балл». */
  output_label?: string;
  eligible: boolean;
  help_id?: string;
}

export interface PredictionSegmentOption {
  id: string;
  human_label: string;
  launchable: boolean;
  help_id?: string;
  date_from?: string | null;
  date_to?: string | null;
  badge?: string | null;
  explanation?: string | null;
  holdout_start?: string | null;
}

export interface StrategyOption {
  id: string;
  technical_id: string;
  human_name: string;
  description: string;
  parameters: Record<string, unknown>;
  help_id?: string;
  frozen?: boolean;
}

export interface CostPresetOption {
  bps: number;
  human_label: string;
  preset: boolean;
  help_id?: string;
}

export interface CostCustomOption {
  allowed: boolean;
  min_bps: number;
  max_bps: number;
  human_label: string;
  help_id?: string;
  friction_model?: string;
}

export interface ResearchDefaults {
  candidate_id: string;
  segment: string;
  policy_id: string;
  risk_id: string;
  commission_bps: number;
  initial_capital: number;
  date_from?: string | null;
  date_to?: string | null;
}

export interface ExecutionAssumptions {
  execution?: string;
  fractional_shares?: boolean;
  dividends?: string;
  benchmark?: string;
  no_leverage?: boolean;
  editable_in_lab?: boolean;
  [key: string]: unknown;
}

export interface ResearchOptions {
  candidates: CandidateOption[];
  prediction_segments: PredictionSegmentOption[];
  policies: StrategyOption[];
  risk_policies: StrategyOption[];
  cost_presets: CostPresetOption[];
  cost_custom: CostCustomOption;
  defaults: ResearchDefaults;
  capital_bounds: { min: number; max: number };
  execution_assumptions: ExecutionAssumptions;
  period_warnings?: { short_calendar_days?: number; min_trading_days_soft?: number };
  quick_suite?: {
    label: string;
    variants: Array<{ policy_id: string; risk_id: string }>;
    costs_bps: number[];
    max_configs: number;
  };
  prediction_ready?: Record<string, unknown> | null;
  holdout_start?: string | null;
}

/** research block appended by backend enrich_run_summary. */
export interface ResearchContext {
  display_name: string;
  note?: string | null;
  created_from?: string | null;
  requested_date_from?: string | null;
  requested_date_to?: string | null;
  observed_holdout: boolean;
  launchable_again: boolean;
  context?: string | null;
  cost_family_fingerprint?: string | null;
}

export interface ResearchSpecSummary extends SimulationSpecSummary {
  risk_name?: string | null;
}

export interface ResearchRunSummary extends Omit<SimulationRunSummary, "spec"> {
  spec?: ResearchSpecSummary | null;
  research?: ResearchContext | null;
  nav_preview?: { points: number; from?: string | null; to?: string | null } | null;
}

export interface LaunchRequest {
  candidate_id: string;
  segment: string;
  policy_id: string;
  risk_id: string;
  commission_bps: number;
  date_from?: string | null;
  date_to?: string | null;
  initial_capital: number;
  name?: string | null;
  note?: string | null;
  force_rerun?: boolean;
}

export type LaunchOutcome = "REUSE_EXISTING" | "CREATED";

export interface LaunchResponse {
  outcome: LaunchOutcome;
  status: string;
  message: string;
  run: ResearchRunSummary;
  config_hash?: string | null;
  cost_family_fingerprint?: string | null;
  warnings?: string[];
  simulation_executed: boolean;
  metrics?: SimulationMetrics | null;
}

export interface CompareDifference {
  field: string;
  human: string;
  values: Record<string, unknown>;
}

export interface CompareMetricRow {
  metric_id: string;
  human_label: string;
  help_id: string;
  values: Record<string, number | null>;
}

export interface CompareNavPoint {
  date: string;
  nav: number;
  drawdown: number;
  nav_normalized: number;
}

export interface CostFamilyCell {
  run_id: number;
  total_price_return?: number | null;
  max_drawdown?: number | null;
  turnover_ratio?: number | null;
  sharpe_rf0?: number | null;
}

export interface CostFamilyRow {
  policy_name?: string | null;
  risk_name?: string | null;
  segment?: string | null;
  date_from?: string | null;
  date_to?: string | null;
  candidate_config_hash?: string | null;
  cells: Record<string, CostFamilyCell>;
}

export interface CostFamily {
  present: boolean;
  message: string;
  matrix?: CostFamilyRow[] | null;
}

export interface CompareResponse {
  runs: ResearchRunSummary[];
  fair_comparison: boolean;
  fair_badge: string;
  /** true when fair and candidate hashes differ (model A/B). */
  model_comparison?: boolean;
  differences: CompareDifference[];
  metrics_table: CompareMetricRow[];
  interpretation: string[];
  observed_holdout_warning?: string | null;
  cost_family: CostFamily;
  nav_series: Record<string, CompareNavPoint[]>;
  normalization?: string;
  period_aligned: boolean;
}

export interface SuiteRequest {
  candidate_id?: string;
  date_from?: string | null;
  date_to?: string | null;
  initial_capital?: number;
}

export interface SuitePlanItem {
  policy_id: string;
  risk_id: string;
  commission_bps: number;
  config_hash: string;
  exists: boolean;
  existing_run_id?: number | null;
}

export interface SuitePlanResponse {
  label: string;
  total: number;
  already_exist: number;
  will_run: number;
  items: SuitePlanItem[];
  not_optimization: boolean;
}

export interface SuiteRunResponse {
  label: string;
  total: number;
  already_exist: number;
  created: number;
  reused: number;
  runs: ResearchRunSummary[];
  not_optimization: boolean;
  message: string;
}

export interface ResearchRunFilters {
  limit?: number;
  policy_id?: string;
  risk_id?: string;
  status?: string;
  segment?: string;
  commission_bps?: number;
  sort?: string;
}

export function getResearchOptions(signal?: AbortSignal): Promise<ResearchOptions> {
  return apiRequest("/research-lab/options", { signal });
}

export async function listResearchRuns(
  filters: ResearchRunFilters = {},
  signal?: AbortSignal,
): Promise<ResearchRunSummary[]> {
  const response = await apiRequest<{ items: ResearchRunSummary[] }>(
    `/research-lab/runs${queryString({
      limit: filters.limit ?? 100,
      policy_id: filters.policy_id,
      risk_id: filters.risk_id,
      status: filters.status,
      segment: filters.segment,
      commission_bps: filters.commission_bps,
      sort: filters.sort,
    })}`,
    { signal },
  );
  return response.items ?? [];
}

export function getResearchRun(
  runId: number | string,
  signal?: AbortSignal,
): Promise<ResearchRunSummary> {
  return apiRequest(`/research-lab/runs/${encodeURIComponent(String(runId))}`, { signal });
}

export function launchResearchRun(
  body: LaunchRequest,
  signal?: AbortSignal,
): Promise<LaunchResponse> {
  return apiRequest("/research-lab/runs", { method: "POST", body, signal });
}

export function compareResearchRuns(
  runIds: Array<number | string>,
  signal?: AbortSignal,
): Promise<CompareResponse> {
  return apiRequest(`/research-lab/compare${queryString({ run_ids: runIds.join(",") })}`, {
    signal,
  });
}

export function planQuickSuite(
  body: SuiteRequest = {},
  signal?: AbortSignal,
): Promise<SuitePlanResponse> {
  return apiRequest("/research-lab/suites/plan", { method: "POST", body, signal });
}

export function runQuickSuite(
  body: SuiteRequest = {},
  signal?: AbortSignal,
): Promise<SuiteRunResponse> {
  return apiRequest("/research-lab/suites", { method: "POST", body, signal });
}
